### PR TITLE
Better validation logic for workflows

### PR DESCRIPTION
**Pull request by mephenor**
This PR introduces a NoOpValidator class that is used as  a dummy to skip validation in a workflow setting.
The new default for transformations is to skip validation if not explicitly instructed to validate.
Validation of input and output can be configured separately.

There's no strong reason for making validation opt-in, apart from the solution being slightly more compact.
If there are strong arguments to keep the validation on by default, the logic can be changed easily.

This also bumps the version to 5.0.0, so we can draft a release based on schemapack.

### CODE REVIEW

File: src/metldata/transform/handling.py [hunk 1]
```diff
@@ -27,6 +34,26 @@
...
+        add_global_plugins: list[type[GlobalValidationPlugin]] | None = None,
+        add_class_plugins: list[type[ClassValidationPlugin]] | None = None,
+        add_resource_plugins: list[type[ResourceValidationPlugin]] | None = None,
+    ):
+        """Do nothing with the input, just skip plugin creation."""
+
+    @override
+    def validate(self, *, datapack: DataPack):
+        """Do nothing and return."""
+        return
```
Comment (sbilge) [inline src/metldata/transform/handling.py:58]: My environment gives `Method "validate" is marked as an override, but no base method was found with this nameMypy[misc](https://mypy.readthedocs.io/en/latest/_refs.html#code-misc)` error.

**Review by sbilge — CHANGES_REQUESTED** [review]
I only noticed a mypy complaint.

Apart from that, I know we currently don’t have negative test cases, but shall we add one here to ensure that a validation error is raised when a datapack is not compatible with its schemapack, both at the beginning and at the end of the process?

Alternatively, we could add an integration test that produces an incompatible intermediate datapack–schemapack pair, although I’m not entirely sure how to simulate that. Perhaps we could use a mock transformation that intentionally creates non-compatible data–model pairs.

File: src/metldata/transform/handling.py (hunk 1 above)
Comment (mephenor) [reply src/metldata/transform/handling.py:58]: Can't reproduce.
Can you check if your environment has the correct mypy version, i.e. rebuild the devcontainer, if necessary?

File: tests/workflow/test_validation.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    PostTransformValidationError,
+    PreTransformValidationError,
+)
+from metldata.workflow.handling import WorkflowHandler
+from tests.fixtures.workflow import VALIDATION_TEST_CASES, WorkflowTestCase
+
+
+@pytest.mark.parametrize("test_case", VALIDATION_TEST_CASES, ids=str)
+def test_workflow_invalid_models(test_case: WorkflowTestCase):
+    """Test the happy path of running a workflow."""
```
Comment (sbilge) [inline tests/workflow/test_validation.py]: Sounds quite unhappy to me :)

File: tests/workflow/test_validation.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        transformation_registry=test_case.transformation_registry,
+        input_model=test_case.input_model,
+    )
+    exception_type = (
+        PreTransformValidationError
+        if test_case.case_name == "invalid_input_model"
+        else PostTransformValidationError
+    )
+    with pytest.raises(exception_type):
+        _ = handler.run(data=test_case.input_data, annotation=test_case.annotation)
```
Comment (sbilge) [inline tests/workflow/test_validation.py]:
```suggestion
        handler.run(data=test_case.input_data, annotation=test_case.annotation)
```

File: tests/fixtures/workflow.py [hunk 4]
```diff
@@ -144,9 +147,21 @@ def _read_test_case(
     )
 
 
+def _read_validation_test_cases() -> list[WorkflowTestCase]:
+    """Read validation test cases."""
+    return [
+        _read_test_case(workflow_dir=WORKFLOW_VALIDATION_DIR, case_name=case_name)
+        for case_name in VALIDATION_WORKFLOWS
```
Comment (sbilge) [inline tests/fixtures/workflow.py:166]: We should create the validation test case differently since those test cases do not include a comparison between the expected and observed transformation results.

File: tests/fixtures/workflow_validation/invalid_input_model/workflow.yaml [hunk 5]
```diff
@@ -0,0 +1,135 @@
+operations:
```
Comment (sbilge) [inline tests/fixtures/workflow_validation/invalid_input_model/workflow.yaml]: Let's switch to a much simpler workflow for this test case

File: tests/fixtures/workflow_validation/invalid_output_model/workflow.yaml [hunk 6]
```diff
@@ -0,0 +1,134 @@
+operations:
```
Comment (sbilge) [inline tests/fixtures/workflow_validation/invalid_output_model/workflow.yaml]: Let's switch to a simpler workflow

File: tests/fixtures/workflow_validation/invalid_input_model/transformed.datapack.yaml
Comment (sbilge) [inline tests/fixtures/workflow_validation/invalid_input_model/transformed.datapack.yaml:1]: We should modify how we create the test cases and delete the unnecessary transformed.* files that do not contribute to the test case.

File: tests/fixtures/workflow_validation/invalid_input_model/transformed.schemapack.yaml
Comment (sbilge) [inline tests/fixtures/workflow_validation/invalid_input_model/transformed.schemapack.yaml:1]: See above

File: tests/fixtures/workflow_validation/invalid_output_model/transformed.datapack.yaml
Comment (sbilge) [inline tests/fixtures/workflow_validation/invalid_output_model/transformed.datapack.yaml:1]: See above

File: tests/fixtures/workflow_validation/invalid_output_model/transformed.schemapack.yaml
Comment (sbilge) [inline tests/fixtures/workflow_validation/invalid_output_model/transformed.schemapack.yaml:1]: See above

File: tests/fixtures/workflow_validation/invalid_input_model/workflow.yaml
Comment (sbilge) [inline tests/fixtures/workflow_validation/invalid_input_model/workflow.yaml:1]: Let's switch to a much simpler workflow

File: tests/fixtures/workflow_validation/invalid_output_model/workflow.yaml
Comment (sbilge) [inline tests/fixtures/workflow_validation/invalid_output_model/workflow.yaml:1]: Let's switch to a much simpler workflow

**Review by sbilge — CHANGES_REQUESTED** [review]

File: tests/workflow/test_validation.py (hunk 3 above)
Comment (mephenor) [reply tests/workflow/test_validation.py]: Wanted to be explicit about this not using the return value, but can simplify it

File: tests/fixtures/workflow.py (hunk 4 above)
Comment (mephenor) [reply tests/fixtures/workflow.py:166]: Done, the transformed_model/transformed_data fields can now be None, skipping loading based on the value of `load_transformed`

File: tests/workflow/test_validation.py (hunk 2 above)
Comment (mephenor) [reply tests/workflow/test_validation.py]: Fixed the desciption

**Comment by mephenor** [comment]
> Alternatively, we could add an integration test that produces an incompatible intermediate datapack–schemapack pair, although I’m not entirely sure how to simulate that. Perhaps we could use a mock transformation that intentionally creates non-compatible data–model pairs.

Added a fixture to intercept constructor call args to check expectations that way.
Added a test case where the second schemapack is invalid to make sure validation only happens at the end.
Added a test case with a workflow containing only one step to test that edge case.

File: tests/workflow/test_validation.py (hunk 3 above)
Comment (sbilge) [reply tests/workflow/test_validation.py]: I think it is clearer that you are not using a return value when you don't assign it to something, because it is already inside of `pytest.raises`

File: tests/workflow/test_validation.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+class CallArgs:
+    """Container to store results from capture_constructor_args"""
+
+    def __init__(self):
+        self.call_args = []
+
+
+@pytest.fixture()
+def capture_constructor_args():
+    """Capture call args for TransformationHandler constructor."""
```
Comment (sbilge) [inline tests/workflow/test_validation.py]: It would be nice to have a bit more description here.

File: tests/workflow/test_validation.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    "invalid_single_step_model": [{"validate_input": True, "validate_output": True}],
+    "invalid_output_model": [
+        {"validate_input": True, "validate_output": False},
+        {"validate_input": False, "validate_output": False},
+        {"validate_input": False, "validate_output": True},
+    ],
+}
+
+
+class CallArgs:
```
Comment (sbilge) [inline tests/workflow/test_validation.py]: How about we call this `ArgsContainer` to prevent `call_args.call_args` repetition in the following fixture?

File: tests/workflow/test_validation.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+        self.call_args = []
+
+
+@pytest.fixture()
+def capture_constructor_args():
+    """Capture call args for TransformationHandler constructor."""
+    call_args = CallArgs()
+
+    def capture(self, *args, **kwargs):
+        call_args.call_args.append(
```
Comment (sbilge) [inline tests/workflow/test_validation.py]: See above

File: tests/workflow/test_validation.py [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+        )
+        self._real_init(*args, **kwargs)
+
+    TransformationHandler._real_init = TransformationHandler.__init__  # type: ignore
+    TransformationHandler.__init__ = capture  # type: ignore
+
+    yield call_args
+
+    TransformationHandler.__init__ = TransformationHandler._real_init  # type: ignore
+    del TransformationHandler._real_init  # type: ignore
```
Comment (sbilge) [inline tests/workflow/test_validation.py:74]: I think pytest's monkeypatch handles saving/restoring automatically. Did you not use it for a reason?

Comment (sbilge) [reply tests/workflow/test_validation.py:74]: Something like this:

```
@pytest.fixture()
def capture_constructor_args(monkeypatch):
    """Capture call args for TransformationHandler constructor."""
    call_args = CallArgs()

    original = TransformationHandler.__init__

    def capture(self, *args, **kwargs):
        call_args.call_args.append(
            {
                "validate_input": kwargs["validate_input"],
                "validate_output": kwargs["validate_output"],
            }
        )
        return original(self, *args, **kwargs)

    monkeypatch.setattr(TransformationHandler, "__init__", capture)

    return call_args
```

**Review by sbilge — CHANGES_REQUESTED** [review]

File: tests/workflow/test_validation.py (hunk 10 above)
Comment (mephenor) [reply tests/workflow/test_validation.py:74]: This will likely overwrite the original (again), based on when exactly the constructor is actually called.
Had that issue with other solution attempts before.
Also, don't really like to use monkeypatch inside a fixture, makes it less transparent.

File: tests/workflow/test_validation.py (hunk 8 above)
Comment (mephenor) [reply tests/workflow/test_validation.py]: Refactored and didn't check... I'll adjust it in a second

File: tests/workflow/test_validation.py (hunk 7 above)
Comment (mephenor) [reply tests/workflow/test_validation.py]: What's missing here?

Comment (mephenor) [reply tests/workflow/test_validation.py]: Added a bit more text.

**Review by sbilge — APPROVED** [review]
